### PR TITLE
iOS TvCasting app: Implemented commissioner discovery and UDC requests

### PR DESCRIPTION
**Pull request by sharadb-amazon**
#### Problem
The iOS TvCasting app does not support commissioner discovery or sending User directed commissioning requests. This PR addresses https://github.com/project-chip/connectedhomeip/issues/14561 and https://github.com/project-chip/connectedhomeip/issues/14560 

#### Change overview
1. New SwiftUI views and their respective view models for CommissionerDiscovery and Commissioning flows.
2. New methods in CastingServerBridge.h/mm to discover commissioners and send UDC requests that talk to CastingServer.cpp (in libTvCastingCommon.a). + CHIP app server initialization in CastingServerBridge.init()
3. Updates to build scripts so that tv-casting-common/CHIPProjectAppConfig.h overrides are picked up. Note the subsequent edit to config/ios/CHIPProjectConfig.h

_TBD: call to openBasicCommissioningWindow or show Onboarding payload will be added in future PR._

#### Testing
* Tested on the iPhone simulator in Xcode against a tv-app running on a RasPi and verified that the TV commissioner is discovered and shown to the user. 
* The user is able to click the button for the discovered commissioner and a UDC request is sent out from the iOS TvCasting app. 
* Screenshots below:

### CODE REVIEW

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm [hunk 1]
```diff
@@ -48,15 +48,72 @@ - (instancetype)init
...
     return self;
 }
 
-// TBD: placeholder will be replaced with true CastingServer functions
-- (int)add:(int)a secondNum:(int)b
+- (bool)discoverCommissioners
 {
-    return a + b;
+    ChipLogProgress(AppServer, "CastingServerBridge().discoverCommissioners() called");
+    CHIP_ERROR err = CastingServer::GetInstance()->DiscoverCommissioners();
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm:74]: Which queue is this running on?  As in, are there going to be thread races here?

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm [hunk 2]
```diff
@@ -48,15 +48,72 @@ - (instancetype)init
...
+    }
+    return true;
 }
 
+- (DiscoveredNodeData *)getDiscoveredCommissioner:(int)index
+{
+    ChipLogProgress(AppServer, "CastingServerBridge().getDiscoveredCommissioner() called");
+
+    DiscoveredNodeData * commissioner = nil;
+    const chip::Dnssd::DiscoveredNodeData * chipDiscoveredNodeData = CastingServer::GetInstance()->GetDiscoveredCommissioner(index);
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm:87]: Again, which queue is this running on?

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm [hunk 3]
```diff
@@ -48,15 +48,72 @@ - (instancetype)init
...
+        ChipLogError(AppServer, "CastingServerBridge().sendUserDirectedCommissioningRequest() failed to parse IP address");
+        return false;
+    }
+
+    chip::Inet::InterfaceId interfaceId = chip::Inet::InterfaceId(platformInterface);
+
+    chip::Transport::PeerAddress commissionerPeerAddress
+        = chip::Transport::PeerAddress::UDP(commissionerAddrInet, commissionerPort, interfaceId);
+
+    CHIP_ERROR err = CastingServer::GetInstance()->SendUserDirectedCommissioningRequest(commissionerPeerAddress);
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm:111]: Same question about queues.

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+@property NSString * hostName;
+
+@property unsigned int platformInterface;
+
+@property NSMutableArray * ipAddresses;
+
+@property size_t numIPs;
+
+- (DiscoveredNodeData *)initWithChipDiscoveredNodeData:(void *)chipDiscoveredNodedata;
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.h:55]: Why is the argument here `void *`?

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.mm [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#import "DiscoveredNodeData.h"
+#include <lib/dnssd/Resolver.h>
+
+@implementation DiscoveredNodeData
+
+- (DiscoveredNodeData *)initWithChipDiscoveredNodeData:(void *)chipDiscoveredNodedata
+{
+    self = [super init];
+    if (self) {
+        chip::Dnssd::DiscoveredNodeData * data = (chip::Dnssd::DiscoveredNodeData *) chipDiscoveredNodedata;
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.mm:29]: We should just have the right type coming in, if possible... This cast is extremely dangerous.

**Review by chrisdecenzo — APPROVED** [review]

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.mm (hunk 5 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.mm]: Done. Avoided the cast by doing the conversion in a separate C++ class: DiscoveredNodeDataConverter

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.h (hunk 4 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/TvCasting/MatterBridge/DiscoveredNodeData.h]: No longer using (void *) pointers. See answer to comment below.

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm (hunk 1 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm]: Updated the code, so this network call and the ones below are made asynchronously on the chipWorkQueue.

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm (hunk 2 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm]: See answer above.

File: examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm (hunk 3 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/TvCasting/MatterBridge/CastingServerBridge.mm]: See answer above.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**